### ISSUE TITLE
UDP endpoint always connected

### ISSUE THREAD

**Issue by roughnecks** [TwiN/gatus#659]
### Describe the bug

I set up a udp endpoint to check ETS2 Dedicated Server. If I stop ETS2 the endpoint still shows the service as connected.

### What do you expect to see?

I'd expect it to fail when the service is not running.

### List the steps that must be taken to reproduce this issue

1. Start ETS2
2. Add this config to config.yaml

```
  - name: euro truck sim 2
    group: game
    url: "udp://IPADDRESS:27016"
    interval: 60s
    conditions:
      - "[CONNECTED] == true"
    alerts:
      - type: email
        description: "healthcheck failed"
        send-on-resolved: true     
```

3. Monitor it, is up
4. stop ETS2
5. monitor it, it's still up for gatus

### Version

GO

**Comment by Katzenbiber** [comment]
I have the same issue with TeamSpeak3. Additionally, the ping times seem unreasonably low with 17ms Europe <-> USA, while normal HTTP trafic takes around 300ms.

**Comment by lefes** [comment]
@TwiN Hello! Can you assign this issue to me? I’ll try to solve it.

**Comment by Katzenbiber** [comment]
![grafik](https://github.com/user-attachments/assets/4dd9aaf3-3155-4cd7-9959-26b35ce3ae41)
I just tested if gatus even sends anything with Wireshark. No wonder there's something wrong, there is literally no packets sent xD

**Comment by Katzenbiber** [comment]
Ok, I tried sending just some random bytes over UDP, but at least Teamspeak doesn't react to it. It strengthens my suspision, that you wouldn't be able to make a generic UDP endpoint tester, as the protocol is different from program to program. 

In theory you could make it send a configurable message which you would expect a response for, and look if the response actually comes back.

**Comment by lefes** [comment]
> Ok, I tried sending just some random bytes over UDP, but at least Teamspeak doesn't react to it. It strengthens my suspision, that you wouldn't be able to make a generic UDP endpoint tester, as the protocol is different from program to program.
> 
> In theory you could make it send a configurable message which you would expect a response for, and look if the response actually comes back.

I agree with you. It seems the UDP endpoint checker is not functioning correctly in this case. A better solution would indeed be to add an option to configure a specific message to check the availability of the service, as you suggested. The default value for this message could be an empty string.

I will work on implementing this feature.

**Comment by lefes** [comment]
@Katzenbiber, I’ve done some more research, and it seems to be a more complex problem. For example, we cannot trigger a UDP port with simple plain text on Teamspeak due to the binary protocol. It seems more logical to monitor these services using other ports (like TCP) or by checking the status of the daemon or containers. 
Implementing a body to check the response doesn’t make any sense. :(

**Comment by Katzenbiber** [comment]
@lefes, thank you for putting in your effort. Yeah, completely agree, either we would have to reverse-engineer every protocol we want to support or make a system flexible enough to let people insert their own protocol into gatus.

@TwiN Do you have any input as to where you want this feature to be going? As far as I can tell, the feature is broken (and hasn't worked probably ever).

**Comment by TwiN** [comment]
> @lefes, thank you for putting in your effort. Yeah, completely agree, either we would have to reverse-engineer every protocol we want to support or make a system flexible enough to let people insert their own protocol into gatus.
> 
> @TwiN Do you have any input as to where you want this feature to be going? As far as I can tell, the feature is broken (and hasn't worked probably ever).

Personally, I would want this feature to work - at least to some extent.

If you want to give it a shot, I'd be more than happy!

**Comment by zImPatrick** [comment]
Has anything changed? I'd love to implement this, however, I'm unsure about where to begin.

I've added a very rudimentary version, which just takes a hex-encoded body from the config, sends that and waits for a response
https://github.com/zImPatrick/gatus/commit/3a9d22b8f72585bd90fb95bc2b23bc9f59240df3#diff-bd3a55a72186f59e2e63efb4951573b2f9e4a7cc98086e922b0859f8ccc1dd09R96-R114

> Implementing a body to check the response doesn’t make any sense. :(
A lot of protocols have handshakes that can just be reused without any problems, so I think it would still be something worth looking into.

For example, to monitor a Counter-Strike server (or most Source Engine games), this could be used:
```yaml
endpoints:
  - name: Counter-Strike server
    url: "udp://194.176.184.32:27015"
    conditions:
      - "[CONNECTED] == true"
    body: FFFFFFFF54536F7572636520456E67696E6520517565727900
    interval: 10s
```

Sadly, it's not very friendly (at least in my eyes) to look at protocol documentation to craft a handshake or go look at one in Wireshark, so I don't know how this could be made any easier

**Comment by TwiN** [comment]
@zImPatrick Nothing has changed so far, and the code from the diff you changed is exactly where the current implementation is (and therefore where to begin). 

Let me know if you have any questions.

**Comment by Katzenbiber** [comment]
@zImPatrick 
> Sadly, it's not very friendly (at least in my eyes) to look at protocol documentation to craft a handshake or go look at one in Wireshark, so I don't know how this could be made any easier

I think the only flexible, but also user friendly way to implement this feature, would be to have presets for the most common usecases, and a generic "body" field for the unusual usecases. This however would basically need a user curated list of the most common udp ping messages for each game server / udp service...

**Comment by jasonshugart** [comment]
I also had a need for UDP monitoring by sending a packet to the endpoint and looking at the response. I agree having some preset values could be useful, but I tackled the implementation using the string escape characters (\x, \u, ...). I implemented this for UDP, TCP, and TLS endpoints. I also needed to get the local address and random strings, so I added template parsing with LocalAddr and RandomString functions. I'll add a PR with my changes.
